drivers/bmx280/bmx280.c

```diff
@@ -23,93 +25,320 @@
-
-#if ENABLE_DEBUG
-static void dump_buffer(const char *txt, uint8_t *buffer, size_t size);
-#define DUMP_BUFFER(txt, buffer, size)  dump_buffer(txt, buffer, size)
+#ifdef BMX280_USE_SPI
+#define BUS                 (dev->params.spi)
+#define CS                  (dev->params.cs)
+#define CLK                 SPI_CLK_5MHZ
+#define MODE                SPI_MODE_0
+#define WRITE_MASK          (0x7F)
 #else
-#define DUMP_BUFFER(txt, buffer, size)
+#define BUS                 (dev->params.i2c_dev)
+#define ADDR                (dev->params.i2c_addr)
 #endif
 
+/* shortcut for accessing byte x of the latest sensor reading */
+#define RAW_DATA            (dev->last_reading)
+
+/* implementation for the driver's configured bus interface (I2C vs SPI) */
+#ifdef BMX280_USE_SPI /* using SPI mode */
+static inline int _acquire(const bmx280_t *dev)
+{
+    if (spi_acquire(BUS, CS, MODE, CLK) != SPI_OK) {
+        return BMX280_ERR_BUS;
+    }
+    return BMX280_OK;
+}
+
+static inline void _release(const bmx280_t *dev)
+{
+    spi_release(BUS);
+}
+
+static int _read_reg(const bmx280_t *dev, uint8_t reg, uint8_t *data)
+{
+    *data = spi_transfer_reg(BUS, CS, reg, 0);
+    return BMX280_OK;
+}
+
+static int _write_reg(const bmx280_t *dev, uint8_t reg, uint8_t data)
+{
+    (void)spi_transfer_reg(BUS, CS, (reg & WRITE_MASK), data);
+    return BMX280_OK;
+}
+
+static int _read_burst(const bmx280_t *dev, uint8_t reg, void *buf, size_t len)
+{
+    spi_transfer_regs(BUS, CS, reg, NULL, buf, len);
+    return BMX280_OK;
+}
+
+#else /* using I2C mode */
+
+static inline int _acquire(const bmx280_t *dev)
+{
+    if (i2c_acquire(BUS) != 0) {
+        return BMX280_ERR_BUS;
+    }
+    return BMX280_OK;
+}
+
+static inline void _release(const bmx280_t *dev)
+{
+    i2c_release(BUS);
+}
+
+static int _read_reg(const bmx280_t *dev, uint8_t reg, uint8_t *data)
+{
+    if (i2c_read_reg(BUS, ADDR, reg, &data, 0) != 1) {
+        return BMX280_ERR_BUS;
+    }
+    return BMX280_OK;
+}
+
+static int _write_reg(const bmx280_t *dev, uint8_t reg, uint8_t data)
+{
+    if (i2c_write_reg(BUS, ADDR, reg, data, 0) != 1) {
+        return BMX280_ERR_BUS;
+    }
+    return BMX280_OK;
+}
+
+static int _read_burst(const bmx280_t *dev, uint8_t reg, void *buf, size_t len)
+{
+    if (i2c_read_regs(BUS, ADDR, reg, buf, (int)len, 0) != (int)len) {
+        return BMX280_ERR_BUS;
+    }
+    return BMX280_OK;
+}
+
+#endif /* bus mode selection */
+
+static uint16_t _to_u16_le(const uint8_t *buffer, size_t offset)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    return (((uint16_t)buffer[offset + 1]) << 8) + buffer[offset];
+#else
+    return (((uint16_t)buffer[offset]) << 8) + buffer[offset + 1];
+#endif
+}
+
+static int16_t _to_i16_le(const uint8_t *buffer, size_t offset)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    return (((int16_t)buffer[offset + 1]) << 8) + buffer[offset];
+#else
+    return (((int16_t)buffer[offset]) << 8) + buffer[offset + 1];
+#endif
+}
+
```

Comment:
We should have this in common code.

Reply:
of course, its long been merged into byteorder.h -> a nice sign of how old this PR has become :-) Will fix.

Reply:
ups, on closer look it seems I was mistaken. These functions do not exist in shared code anywhere until now. IMO they are borderline in terms of being generic enough to be moved into `byteorder.h` or similar or just stay in this module. I'd like to merge them as is to take some more time to think about what better to do here...

Reply:
Functions to read from an unaligned  buffer are always useful.